### PR TITLE
Made talent cards use the same renderer as spells.

### PR DESCRIPTION
**Pull request by ThatRanga**
There's one bug that I can't figure out why it's happening, any insight could be great. Only one card ends up like this and can't tell what in the data causes this.

image

### CODE REVIEW

File: src/talents.csv [hunk 1]
```diff
@@ -1,23 +1,23 @@
 "Name","Class","Level","Raw Description","Icon Link"
-"Feral Agression","Druid","4","When you use <b>Demoralizing Roar</b>, add an additional {dice red 1} {dice green 1} to dice pool.<br> <b>Ferocious Bite</b> instead allows you to change one {dice red,blue 1} into a {dice red 1 8} for each Energy you spend.","https://wow.zamimg.com/images/wow/icons/large/classic_ability_druid_demoralizingroar.jpg"
-"Nature's Grasp","Druid","3","<i>End of your Dice Pool step:</i> You may reroll any number of {dice green 1} once.<br> These dice may also be rerolled during your Reroll step using your Reroll value.","https://wow.zamimg.com/images/wow/icons/large/spell_nature_natureswrath.jpg"
+"Feral Agression","Druid","4","When you use <b>Demoralizing Roar</b>, add an additional {dice red 1} {dice green 1} to dice pool.</p><b>Ferocious Bite</b> instead allows you to change one {dice red,blue 1} into a {dice red 1 8} for each Energy you spend.","https://wow.zamimg.com/images/wow/icons/large/classic_ability_druid_demoralizingroar.jpg"
```
Comment (rissole) [inline src/talents.csv]: Feral Aggression*

**Review by rissole — APPROVED** [review]
i should probably rewrite the parser thing. i like bens recursive idea - then in theory we can support nested tags too.

does the bug still happen if you change line 34 in descriptionParser to `currentConstructingNodeIndex += 2` instead of +1?

**Comment by bqx** [comment]
> There's one bug that I can't figure out why it's happening, any insight could be great. Only one card ends up like this and can't tell what in the data causes this.


I see a few more cards displaying Object object, probably something in the parser that doesn't have to be fixed in this PR

**Review by bqx — APPROVED** [review]

File: src/components/CardTalent/index.tsx [hunk 2]
```diff
@@ -1,11 +1,30 @@
...
+    <TalentCarouselCard>
+      <h1 style={{ color: COLORS.foregroundPrimary }}>{talent.name}</h1>
+      <IconWrapper>
+        <Icon path={talent.iconLink} width={64} />
+      </IconWrapper>
+      <BasicInfoContainer>
+        <div>Level {talent.requiredLevel}</div>
+      </BasicInfoContainer>
+      <DescriptionContainer>{memoizedDescription}</DescriptionContainer>
+    </TalentCarouselCard>
```
Comment (bqx) [inline src/components/CardTalent/index.tsx:24]: /nit: There's some box-shadow and border-radius on the PowerCarouselCard that would be good to bring across here (or ideally re-use)

File: src/util/descriptionParser.tsx [hunk 3]
```diff
@@ -0,0 +1,47 @@
+import { createElement, ReactNode } from "react";
+
+const RAW_DESCRPTION_TAG_NAME_MAP: { [k: string]: string } = { b: "strong", i: "em" };
```
Comment (bqx) [inline src/util/descriptionParser.tsx]: /nit: typo in "DESCRIPTION"

File: src/util/descriptionParser.tsx [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+
+  for (let i = processedString.length; i < stringLeftToProcess.length; i++) {
+    processedString += stringLeftToProcess[i];
+    const stringChunk = processedString.slice(processedString.length - 3);
+
+    //See if we're dealing with a tag, if not continue to process string
+    if (stringChunk.startsWith("</") || (stringChunk.startsWith("<") && stringChunk.endsWith(">"))) {
+      //We're dealing with luke's crazy paragraph things
+      if (stringChunk === "</p") {
+        nodes.push(processedString.slice(0, processedString.length - 3), createElement("p"));
```
Comment (bqx) [inline src/util/descriptionParser.tsx]: Is this correct? I think we are creating `p` nodes without any content in them. Edit: if this is the way it is generated currently I think its fine to merge as-is and maybe fix up later

File: src/components/CarouselClassTalentsSelect/index.tsx [hunk 5]
```diff
@@ -15,7 +15,7 @@ export default function SelectTalents(props: Props) {
     .filter((talent) => !props.equippedTalents.includes(talent.name))
     .map((talent) => ({
       name: talent.name,
-      renderNode: () => <CardTalent title={talent.name} description={talent.rawDescription} />,
+      renderNode: () => <CardTalent talent={talent} key={talent.name} />,
```
Comment (bqx) [inline src/components/CarouselClassTalentsSelect/index.tsx:18]: /nit: I don't think this needs a `key`, unless Talents are rendered differently - these get wrapped in an element in the Carousel which has the key. Ignore this if I'm wrong  :\

File: src/components/CardTalent/index.tsx [hunk 6]
```diff
@@ -14,13 +33,26 @@ const TalentCarouselCard = styled.div`
...
+
+const IconWrapper = styled.div`
+  display: flex;
+  justify-content: center;
+  margin-bottom: 8px;
+`;
+
+const DescriptionContainer = styled.div`
+  flex: 1;
+`;
```
Comment (bqx) [inline src/components/CardTalent/index.tsx:56]: Doesn't have to be in this PR, but we should probs share these styled components when we introduce Item cards

**Review by bqx — APPROVED** [review]

File: src/util/descriptionParser.tsx [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+  let stringLeftToProcess: string = rawDescription;
+  let processedString: string = "";
+
+  for (let i = processedString.length; i < stringLeftToProcess.length; i++) {
+    processedString += stringLeftToProcess[i];
+    const stringChunk = processedString.slice(processedString.length - 3);
+
+    //See if we're dealing with a tag, if not continue to process string
+    if (stringChunk.startsWith("</") || (stringChunk.startsWith("<") && stringChunk.endsWith(">"))) {
+      //We're dealing with luke's crazy paragraph things
```
Comment (rissole) [inline src/util/descriptionParser.tsx:23]: glad you maintained the statement on your sanity

File: src/util/descriptionParser.tsx [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+  nodes: ReactNode[];
+  stringLeft: string;
+} {
+  const nodes: ReactNode[] = [];
+  let stringLeftToProcess: string = rawDescription;
+  let processedString: string = "";
+
+  for (let i = processedString.length; i < stringLeftToProcess.length; i++) {
+    processedString += stringLeftToProcess[i];
+    const stringChunk = processedString.slice(processedString.length - 3);
```
Comment (rissole) [inline src/util/descriptionParser.tsx:19]: the fact that this slice goes negative fucks with my head.

how extensible will this be to handle the `{dice ...}` case? it looks pretty hard coded to chunks of 3

i think there's a neater way to do this but it needs some thought

**Review by rissole — APPROVED** [review]
happy with this cause it works for now we can worry about `{dice` when we come to it

File: src/util/descriptionParser.tsx (hunk 8 above)
Comment (bqx) [reply src/util/descriptionParser.tsx:19]: Isn't this fine? Its just pulling in characters until it ends with a tag, we can just match for {dice ...} with a regex and do the same thing as here

Comment (rissole) [reply src/util/descriptionParser.tsx:19]: yeah it's fine

**Pull request merged**